scene/animation/animation_node_state_machine.cpp

```diff
@@ -998,7 +1030,10 @@ bool AnimationNodeStateMachinePlayback::_can_transition_to_next(AnimationTree *p
 				playback = playback->duplicate();
 			}
 			playback->_next_main();
-			// Then, fadeing should be end.
+			// Then, fading should end.
+			if (!fading_from.is_empty() && !_is_grouped_state_machine(p_state_machine->get_node(fading_from))) {
+				emit_signal(SceneStringName(node_finished), fading_from);
+			}
```

Comment:
It seems to me that the places to fire signals are scattered all over the place. Can you make a method to assign a state, for example, `fading_from = “”` and unify them there?